### Term Deposit Subscription Prediction :

#### Import Necessary Libraries :

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

#### Read Files :

In [2]:
df = pd.read_csv(r"G:\Imarticus LMS\DT And RF Using Python Part 2\bank.csv")

#### View all rows and columns :

In [3]:
pd.options.display.max_columns = None
pd.options.display.max_rows = None

In [4]:
df.head()

age         job  marital  education default  balance housing loan  contact  \
0   59      admin.  married  secondary      no     2343     yes   no  unknown   
1   56      admin.  married  secondary      no       45      no   no  unknown   
2   41  technician  married  secondary      no     1270     yes   no  unknown   
3   55    services  married  secondary      no     2476     yes   no  unknown   
4   54      admin.  married   tertiary      no      184      no   no  unknown   

   day month  duration  campaign  pdays  previous poutcome deposit  
0    5   may      1042         1     -1         0  unknown     yes  
1    5   may      1467         1     -1         0  unknown     yes  
2    5   may      1389         1     -1         0  unknown     yes  
3    5   may       579         1     -1         0  unknown     yes  
4    5   may       673         2     -1         0  unknown     yes

#### Display Column Names :


In [5]:
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'balance', 'housing',
       'loan', 'contact', 'day', 'month', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'deposit'],
      dtype='object')

####  Get the number of Rows and Columns :


In [6]:
df.shape

(11162, 17)

#### Check for any misssing null values in the data :


In [7]:
df.isna().sum()

age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
deposit      0
dtype: int64

##### Get Datatypes :

In [8]:
df.dtypes

age           int64
job          object
marital      object
education    object
default      object
balance       int64
housing      object
loan         object
contact      object
day           int64
month        object
duration      int64
campaign      int64
pdays         int64
previous      int64
poutcome     object
deposit      object
dtype: object

##### Transform The Data :

##### Transform non-numerical into numerical column :

In [9]:
from sklearn.preprocessing import LabelEncoder
for column in df.columns:
        if df[column].dtype == np.number :
            continue
        df[column] = LabelEncoder().fit_transform(df[column])

In [10]:
df.head()

age  job  marital  education  default  balance  housing  loan  contact  \
0   41    0        1          1        0     2288        1     0        2   
1   38    0        1          1        0      469        0     0        2   
2   23    9        1          1        0     1618        1     0        2   
3   37    7        1          1        0     2356        1     0        2   
4   36    0        1          2        0      608        0     0        2   

   day  month  duration  campaign  pdays  previous  poutcome  deposit  
0    4      8      1021         0      0         0         3        1  
1    4      8      1283         0      0         0         3        1  
2    4      8      1246         0      0         0         3        1  
3    4      8       577         0      0         0         3        1  
4    4      8       671         1      0         0         3        1

##### Drop the age column :

In [11]:
df = df.drop('age' , axis = 1)

In [12]:
df.head()

job  marital  education  default  balance  housing  loan  contact  day  \
0    0        1          1        0     2288        1     0        2    4   
1    0        1          1        0      469        0     0        2    4   
2    9        1          1        0     1618        1     0        2    4   
3    7        1          1        0     2356        1     0        2    4   
4    0        1          2        0      608        0     0        2    4   

   month  duration  campaign  pdays  previous  poutcome  deposit  
0      8      1021         0      0         0         3        1  
1      8      1283         0      0         0         3        1  
2      8      1246         0      0         0         3        1  
3      8       577         0      0         0         3        1  
4      8       671         1      0         0         3        1

#### Transform The Data :

##### Transform non-numerical into numerical column :

In [13]:
from sklearn.preprocessing import LabelEncoder
for column in df.columns:
        if df[column].dtype == np.number :
            continue
        df[column] = LabelEncoder().fit_transform(df[column])

In [14]:
df.head()

job  marital  education  default  balance  housing  loan  contact  day  \
0    0        1          1        0     2288        1     0        2    4   
1    0        1          1        0      469        0     0        2    4   
2    9        1          1        0     1618        1     0        2    4   
3    7        1          1        0     2356        1     0        2    4   
4    0        1          2        0      608        0     0        2    4   

   month  duration  campaign  pdays  previous  poutcome  deposit  
0      8      1021         0      0         0         3        1  
1      8      1283         0      0         0         3        1  
2      8      1246         0      0         0         3        1  
3      8       577         0      0         0         3        1  
4      8       671         1      0         0         3        1

#### Again get DataTypes :

In [15]:
df.dtypes

job          int64
marital      int64
education    int64
default      int64
balance      int64
housing      int64
loan         int64
contact      int64
day          int64
month        int64
duration     int64
campaign     int64
pdays        int64
previous     int64
poutcome     int64
deposit      int64
dtype: object

#### Split the Data into train and test :

In [16]:
from sklearn.model_selection import train_test_split

In [17]:
df_train , df_test = train_test_split(df , test_size = .2 ,random_state = 555)

In [18]:
df_train_x = df_train.iloc[: , 0:-1]
df_train_y = df_train.iloc[: , -1]

In [19]:
df_test_x = df_test.iloc[: , 0:-1]
df_test_y = df_test.iloc[: , -1]

In [20]:
print(df_train_x.shape)
print(df_train_y.shape)

(8929, 15)
(8929,)


In [21]:
print(df_test_x.shape)
print(df_test_y.shape)

(2233, 15)
(2233,)


#### Now Apply Decision Tree Classifier  :

#### Model Formation :

In [22]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier()
dt.fit(df_train_x , df_train_y)

DecisionTreeClassifier()

#### Prediction :

In [23]:
pred_dt  = dt.predict(df_test_x )
pred_dt

array([0, 1, 0, ..., 1, 1, 1], dtype=int64)

#### Confusion Matrix :

In [24]:
from sklearn.metrics import confusion_matrix
tab_dt =  confusion_matrix(pred_dt   , df_test_y )
tab_dt

array([[952, 272],
       [223, 786]], dtype=int64)

#### Efficiency :

In [25]:
tab_dt.diagonal().sum() * 100  / tab_dt.sum()


77.83251231527093

# .......................................................................................................

#### Now Apply Random Forest :


#### Model Formation :

In [26]:
from sklearn.ensemble import RandomForestClassifier
rfc  = RandomForestClassifier()


In [27]:
rfc  = RandomForestClassifier(n_estimators  = 10)
rfc.fit(df_train_x , df_train_y)

RandomForestClassifier(n_estimators=10)

#### Prediction :

In [28]:
pred_rfc  = rfc.predict(df_test_x )
pred_rfc

array([0, 1, 0, ..., 1, 1, 1], dtype=int64)

#### Confusion Matrix :

In [29]:
tab1_dt  = confusion_matrix(pred_rfc , df_test_y)
tab1_dt

array([[1009,  219],
       [ 166,  839]], dtype=int64)

#### Efficiency :

In [30]:
tab1_dt.diagonal().sum() * 100  / tab1_dt.sum()

82.75862068965517

# .............................................................................................................

#### Now Apply ADA Boost :

#### Model Formation :


In [31]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier


In [32]:
dt = DecisionTreeClassifier()

In [33]:
abc = AdaBoostClassifier( dt   ,  10)


In [34]:
abc.fit(df_train_x , df_train_y)


AdaBoostClassifier(base_estimator=DecisionTreeClassifier(), n_estimators=10)

#### Prediction :


In [35]:
pred_abc = abc.predict(df_test_x)
pred_abc

array([0, 1, 0, ..., 1, 1, 1], dtype=int64)

#### Confusion Matrix :


In [36]:
tab_abc = confusion_matrix(pred_abc , df_test_y)
tab_abc

array([[959, 263],
       [216, 795]], dtype=int64)

#### Efficiency :

In [37]:
tab_abc.diagonal().sum() * 100  / tab_abc.sum()

78.54903716972683

# ..........................................................................................

### Conclusion :

#### Here I have built various types of model such as Decision Tree Classifier , Random Forest Classifier and ADA Boost Model .

### Accuracy / Efficiency :

#### Decision Tree Classifier         =  77.83251231527093

#### Random Forest Classifier      =  82.75862068965517

#### ADA Boost                               =  78.54903716972683


#### By cmparing above Accuracies / Efficiencies , I have selected Random Forest Classifier for Model Formation , because it has highest Accuracy , Which is equal to  82.75862068965517 .

# ................................. END.......................................................